### ISSUE TITLE
x/net/http2: support closure of all connections and graceful shutdown

### ISSUE THREAD

**Issue by nhooyr** [golang/go#26682]
Very happy to see h2c supported: https://github.com/golang/net/commit/c4299a1a0d8524c11563db160fbf9bddbceadb21

However, there is no way to enable graceful shutdown, or even close all h2c connections. We should support this.

You could use `http2.ConfigureServer` to configure a http2 server's graceful shutdown to be called when the http server is shutdown but the http server's shutdown method will not wait for the http2 connections to close because they have been hijacked.

I think it'd be simplest if we added a `Shutdown` and `Close` method to `*http.Server` analogous to `*http.Server`'s methods.

**Comment by obeattie** [comment]
Is there a workaround possible for this? As far as I can see it's completely impossible to do at present, as the methods that need to be invoked to gracefully shutdown a connection (eg. `serverConn.startGracefulShutdown`) are unexported.

**Comment by obeattie** [comment]
Is there a workaround possible for this? As far as I can see it's completely impossible to do at present, as the methods that need to be invoked to gracefully shutdown a connection (eg. `serverConn.startGracefulShutdown`) are unexported. cc. @bradfitz

**Comment by nhooyr** [comment]
That does not work as the Shutdown itself doesn't wait for the h2c connections.

See https://github.com/golang/go/blob/86ce1cb060446bcee2454080c7f55e754dbf7f82/src/net/http/server.go#L2631

It starts the graceful shutdown of h2c connections in a new goroutine but only wait for the unhijacked connections.

**Comment by bradfitz** [comment]
Sorry, h2c is super low on my priority list. I don't have time to investigate this enough to be helpful here.

**Comment by akshayjshah** [comment]
Would the Go team be open to CLs that address this? (I see that there's a very old CL open, but it doesn't seem to have been reviewed.) h2c is particularly popular in the gRPC ecosystem, and it'd be nice for gRPC implementations built on `net/http` to support graceful shutdown. We've had a few [Connect](https://github.com/bufbuild/connect-go) users complain about this 😢

Is @neild the right person to take a look at `x/net/http` and `net/http` issues?

**Comment by sword-jin** [comment]
Hi, how's this issue going so far?
will we take this problem into go stdlib? refer https://github.com/golang/go/issues/67810

**Comment by aimuz** [comment]
Is there any temporary fix for this?

**Comment by neild** [comment]
Go 1.24 contains direct support for unencrypted HTTP/2 connections in net/http. See https://pkg.go.dev/net/http#example-Protocols-Http1, and https://pkg.go.dev/net/http#Protocols.SetUnencryptedHTTP2.

When configuring HTTP/2 in this way, `Server.Shutdown` and `Server.Close` should work for unencrypted HTTP/2 connections as for any other.

**Comment by aimuz** [comment]
@neild

When I use h2c to provide the GRPC protocol (connectrpc), calling Shutdown does not wait for the client to close the connection, and the client encounters a socket closed error.

The grpc server is also based on HTTP2, and I would like to know if there will be any inconsistencies here?


```go
func (s *Server) Start(ctx context.Context, addr string) error {
	protocols := &http.Protocols{}
	protocols.SetHTTP1(true)
	protocols.SetUnencryptedHTTP2(true)

	server := http.Server{
		Addr:      addr,
		Handler:   withCORS(s.mux),
		Protocols: protocols,
	}

	errCh := make(chan error, 1)
	go func() {
		s.log.Info("Starting ConnectRPC server", "addr", addr)
		err := server.ListenAndServe()
		if errors.Is(err, http.ErrServerClosed) || err == nil {
			return
		}
		errCh <- err
	}()

	select {
	case <-ctx.Done():
		s.log.Info("Shutting down the ConnectRPC server")
		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
		defer cancel()
		if err := server.Shutdown(ctx); err != nil {
			s.log.Error("Failed to shutdown ConnectRPC server", "err", err)
		}
	case err := <-errCh:
		return err
	}
	return nil
}
```

**Comment by kaiburjack** [comment]
Yeah, currently the Go h2c implementation sends only a single non-graceful GOAWAY which might race with the client sending a new request.
The way how this should be done gracefully is also *explicitly* pointed-out in the [HTTP/2 RFC 9113 chapter 6.8 "GOAWAY"](https://datatracker.ietf.org/doc/html/rfc9113#:~:text=A%20client%20that%20is%20unable%20to%20retry,be%20cleanly%20shut%20down%20without%20losing%20requests.):

> A client that is unable to retry requests loses all requests that are
> in flight when the server closes the connection. This is especially
> true for intermediaries that might not be serving clients using
> HTTP/2. A server that is attempting to gracefully shut down a
> connection SHOULD send an initial GOAWAY frame with the last stream
> identifier set to 231-1 and a NO_ERROR code. This signals to the
> client that a shutdown is imminent and that initiating further
> requests is prohibited. After allowing time for any in-flight stream
> creation (at least one round-trip time), the server MAY send another
> GOAWAY frame with an updated last stream identifier. This ensures
> that a connection can be cleanly shut down without losing requests.

Some HTTP/2 servers (best example is the Envoy proxy) implement this correctly and allow for this race to be resolved by sending the double-GOAWAY pattern with a 5s (default) delay between the first graceful GOAWAY (with max stream id) and the final GOAWAY (with last seen stream id).
Currently, Go's implementation is not graceful, but it could be, given that HTTP/2 is able to resolve this race which was a huge and inherently unresolvable problem with HTTP/1.1 because in HTTP/1.1 there was _no_ in-band way to signal to the client that the connection is about to be closed (other than actually closing it or providing the client with out-of-band information about the server closing soon).
